lib/Conversion/TorchToTcp/Elementwise.cpp

```diff
@@ -60,54 +60,144 @@ class ConvertAtenAddSubOp : public OpConversionPattern<AtenOpT> {
     RankedTensorType lhsType = lhs.getType().dyn_cast<RankedTensorType>();
 
     Value rhs = adaptor.getOther();
-    RankedTensorType rhsType = rhs.getType().dyn_cast<RankedTensorType>();
 
     RankedTensorType resultType =
         OpConversionPattern<AtenOpT>::getTypeConverter()
             ->convertType(op.getType())
             .template cast<RankedTensorType>();
 
-    if (!lhsType || !rhsType || !resultType)
+    if (!lhsType || !resultType)
       return rewriter.notifyMatchFailure(
           op, "Only Ranked Tensor types are supported in TCP");
 
-    lhs = torch_to_tcp::broadcastInLeadingDimsToMatchShape(rewriter, lhs, rhs);
-    rhs = torch_to_tcp::broadcastInLeadingDimsToMatchShape(rewriter, rhs, lhs);
+    auto inputAType = op.getSelf()
+                          .getType()
+                          .template dyn_cast<torch::Torch::ValueTensorType>()
+                          .getDtype();
+    auto outputType = op.getType()
+                          .template dyn_cast<torch::Torch::ValueTensorType>()
+                          .getDtype();
+
+    if (isa<AtenAddScalarOp>(op) || isa<AtenSubScalarOp>(op)) {
+      RankedTensorType tensorResultType =
+          RankedTensorType::get({}, adaptor.getOther().getType());
+      rhs = torch_to_tcp::scalarToTcpTensor(rewriter, op, tensorResultType,
```

Comment:
We should move `scalarToTcpTensor` to be its own `OpConversionPattern` that converts 
`torch.const.int -> tcp.const` and just use `op.getOther`

Reply:
Currently, `tcp.const` only outputs `RankedTensorType`. For ops that still requires the scalar input, always converting `torch.const.int -> tcp.const` will cause problems in the default `backendTypeConversion` pass due to the type mismatch. For elementwise ops with the scalar input,  we use `getConstTensor` utils which creates a `tcp::ConstOp` to get the const tensor first, then cast and broadcast its type and shape. For other places which still requires the `torch.const` input, it will be automatically converted to `arith.const` op. Please let me know if this makes any sense or I misunderstand something.

Reply:
Just to make sure I understand correctly: which cases we don't want to lower `torch.const.int -> tcp.const` ?

Reply:
I guess any op that still requires scalar (torch.const) as its input will have this problem. I tried to implement the `torch.const-> tcp.const` conversion pattern for both int and fp, but fails bunch of e2e tests. For example, if we convert `torch.const.int -> tcp.const`, `aten.add.scalar` (which requires a scalar as its second input) will produce following error: BackendTypeConversion.cpp:118: auto setupTorchFloatToF64Conversion(mlir::ConversionTarget &, mlir::TypeConverter &)::(anonymous class)::operator()(mlir::OpBuilder &, mlir::Float64Type, mlir::ValueRange, mlir::Location) const: Assertion `inputs[0].getType().isa<Torch::FloatType>()' failed.

Reply:
Aren't we converting `aten.add.scalar -> tcp.add` and broadcasting the scalar to same shape as the lhs ?

Reply:
I agree with @zezhang's point that converting all `torch.const-> tcp.const` would not work because `tcp.const` always results in a tensor. 

This would be a problem for any op that takes in a scalar. For example, `tcp.broadcast` op takes in broadcast sizes as a Variadic list of scalars (https://github.com/llvm/torch-mlir/blob/mlir-tcp/externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td#L202). If they are constants, we can't be using `tcp.const` to return `index` types (at least with the current specification of `tcp.const`).

Are you suggesting we change `tcp.const` to allow returning scalars as well?

Reply:
> Are you suggesting we change tcp.const to allow returning scalars as well?

Yes, I can see use writing less conversion pattern code if we allow `tcp.const` to return scalar or tensor types e.g
```mlir
%0 = tcp.constant 2 : f32
%1 = tensor.from_ements %0 : tensor<f32>
```
will add folding pattern to do:
```mlir
%0 = tcp.constant 2 : tensor<f32>
```
and so on, otherwise every op we will need to have handle scalar operand case.

Reply:
Agree that it would help reduce the conversion pattern code. But can we base design decisions on that factor?

So far, all TCP ops return only Tensors, none of them produce scalars. My worry is that it might be a bigger change than just updating `tcp.const`, which might have other design implications. We should probably carefully consider this and do it in a separate PR, if there is an agreement. Wdyt?

Reply:
> But can we base design decisions on that factor?

My concern is current design converting  `TorchOp -> TCPOp` requires  special handling of `TCPOp` based on if operands are scalar or tensors seems like a gap we have in the `TCP` dialect and how it models constants.

>We should probably carefully consider this and do it in a separate PR, if there is an agreement. Wdyt?

Yes, my comment here is more around understanding current Torch -> TCP approach so I can provide useful feedback on this PR its nonblocking indeed we shouldn't address design changes here!